### PR TITLE
Merge gsl/develop to RRFS_dev

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
# PR Checklist

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [ ] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR 
are specified below.

- [x] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below. **Changes to code from many people past many moons.**

- [x] New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.

## Description

Merge gsl/develop to RRFS_dev

### Issue(s) addressed

Merge gsl/develop to RRFS_dev

## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

**Testing is incomplete**

- [x] hera.intel
- [x] hera.gnu
- [ ] orion.intel
- [ ] cheyenne.intel 
- [ ] cheyenne.gnu
- [ ] gaea.intel 
- [x] jet.intel
- [ ] wcoss_cray
- [ ] wcoss_dell_p3
- [ ] opnReqTest for newly added/changed feature
- [ ] CI

## Dependencies

https://github.com/NOAA-GSL/fv3atm/pull/149
https://github.com/NOAA-GSL/ccpp-physics/pull/157

### CODE REVIEW

**Review by hannahcbarnes — APPROVED** [review]

**Review by mdtoyNOAA — APPROVED** [review]

File: tests/parm/diag_table/diag_table_hrrr [hunk 1]
```diff
@@ -282,28 +282,20 @@
...
 "gfs_phys",    "nddsf_ave",    "nddsf_ave",     "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "trans_ave",    "transp_ave",    "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evbs_ave",     "direvap_ave",   "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evcw_ave",     "canevap_ave",   "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "sbsno",        "sublim",        "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evbs",         "direvap",       "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evcw",         "canevap",       "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "trans",        "transp",        "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "snowmt_land",  "snom_land",     "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "snowmt_ice",   "snom_ice",      "fv3_history2d",  "all",  .false.,  "none",  2
```
Comment (tanyasmirnova) [inline tests/parm/diag_table/diag_table_hrrr]: @SamuelTrahanNOAA Sam, I would like to keep the RRFS_dev version of diag_table_hrrr in the merged version.

File: tests/parm/diag_table/diag_table_hrrr_stoch [hunk 2]
```diff
@@ -281,28 +282,20 @@
...
-"gfs_phys",    "trans_ave",    "transp_ave",    "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evbs_ave",     "direvap_ave",   "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evcw_ave",     "canevap_ave",   "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "sbsno",        "sublim",        "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evbs",         "direvap",       "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "evcw",         "canevap",       "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "trans",        "transp",        "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "snowmt_land",  "snom_land",     "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_phys",    "snowmt_ice",   "snom_ice",      "fv3_history2d",  "all",  .false.,  "none",  2
+"gfs_phys",    "trans_ave",    "trans_ave",     "fv3_history2d",  "all",  .false.,  "none",  2
```
Comment (tanyasmirnova) [inline tests/parm/diag_table/diag_table_hrrr_stoch]: same for the the diag_table_hrrr_stoch. The version in RRFS_dev has changes specific for the use of RUC LSM.

**Review by tanyasmirnova — COMMENTED** [review]
@SamuelTrahanNOAA @hu5970 Ming, I think we have to run a test case with the merged version to make sure that we do not substantially change the result.

**Comment by SamuelTrahanNOAA** [comment]
This is a few months of development. I'd expect a noticeable change in results.

**Comment by hu5970** [comment]
I will test this branch using RRFS for couple of cycles. If everything runs smoothly, we can merge this and start a retro test.

**Comment by SamuelTrahanNOAA** [comment]
> I will test this branch using RRFS for couple of cycles. If everything runs smoothly, we can merge this and start a retro test.

You should wait until I finish the regression tests.

**Comment by SamuelTrahanNOAA** [comment]
The update broke the netcdf io. I'm working on fixing it now.

**Comment by tanyasmirnova** [comment]
@hu5970 Ming, will you save the real-time cycles with the old code for comparison?

**Comment by hu5970** [comment]
@tanyasmirnova  I am running two short retros on Jet to test the new code:

1) The retro with the new model code:
/mnt/lfs4/BMC/rtwbl/mhu/rrfs/baseE/NCO_dirs/stmp/tmpnwprd/RRFS_conus_3km
run from 03Z to 08z spin-up cycle and 09Z-15Z product cycles.

2) the retro with current real-time model code:
/mnt/lfs4/BMC/rtwbl/mhu/rrfs/testE/NCO_dirs/stmp/tmpnwprd/RRFS_conus_3km
run from 03Z to 08z spin-up cycle and 09Z product cycles.

For retro 2, I also tested the new chgres_cube and the new chgres_cube can product similar results as the current one.

So, please check compare retro 1 and retro 2. If the results look reasonable, I will suggest to merge the PR. 
I will update the APP hash and add new thompson code as a new BaseE and start a long retro on Hera.

Thanks,
Ming

**Comment by SamuelTrahanNOAA** [comment]
You can run retros without the netcdf fix, but it'll run slower.

**Comment by tanyasmirnova** [comment]
@hu5970 Ming, the retro with the new code does not use updated diag_table that has some extra output variables, and also has removed some unnecessary variables.

**Comment by hu5970** [comment]
@tanyasmirnova Please check:
/mnt/lfs4/BMC/rtwbl/mhu/rrfs/baseE/ufs-srweather-app/src/ufs_weather_model
and let me know how to update the diag_table. 
We can try with new diag_table with this retro for now.

Also, is temperature/wind/moisture fields look OK?

**Comment by hu5970** [comment]
@SamuelTrahanNOAA Yes, warm start runs slow but they run without any problem.

**Comment by tanyasmirnova** [comment]
@hu5970 Ming, the diag_table_hrrr in the /mnt/lfs4/BMC/rtwbl/mhu/rrfs/baseE/ufs-srweather-app/src/ufs_weather_model is not what is in the retro. 
Please, check /mnt/lfs4/BMC/wrfruc/smirnova/RRFS_lake/2021072618/diag_table for reference. The diag_table_hrrr in /mnt/lfs4/BMC/rtwbl/mhu/rrfs/baseE/ufs-srweather-app/src/ufs_weather_model is very close to reference table, just one variable is missing.
I will follow up on the results later.

**Comment by SamuelTrahanNOAA** [comment]
@tanyasmirnova If you want new diag_tables in this PR, let me know.

**Comment by tanyasmirnova** [comment]
@SamuelTrahanNOAA Sam, I would like to add diag_table_hrrr to this PR, it should be as in RRFS_dev branch.

**Comment by tanyasmirnova** [comment]
@hu5970 @joeolson42 @mdtoyNOAA 
Ming,
I checked the 1-h fcst in the first spin-up cycle hoping for very close results.
I see the larges differences in the wind at the low levels:

Screen Shot 2022-06-16 at 1 19 58 PM
Screen Shot 2022-06-16 at 1 19 08 PM
The same to 10-m wind. Other surface variables are close. The 2-m temperature and specific humidity are different as expected ad the new code has HRRR 2-m diagnostics. Skin temperature differs in certain regions:
Screen Shot 2022-06-16 at 1 35 57 PM
Looks like the skin temperature diffs are related to the wind diffs.

**Comment by SamuelTrahanNOAA** [comment]
> Sam, I would like to add diag_table_hrrr to this PR, it should be as in RRFS_dev branch.

I already made that change last night.

**Comment by tanyasmirnova** [comment]
@SamuelTrahanNOAA Then what did you refer to in your comment: "If you want new diag_tables in this PR, let me know"

**Comment by SamuelTrahanNOAA** [comment]
I meant, let me know if you want me to change any diag_tables in this branch. The two you asked me to revert now match the top of RRFS_dev.

**Comment by hu5970** [comment]
@tanyasmirnova The diag_table used for model run is from regional workflow:

 /mnt/lfs4/BMC/rtwbl/mhu/rrfs/baseE/ufs-srweather-app/regional_workflow/ush/templates/diag_table.FV3_HRRR

Do you think we should replace the above diag table with your:
/mnt/lfs4/BMC/wrfruc/smirnova/RRFS_retro/gsl-srweather-app/regional_workflow/ush/templates//diag_table.FV3_HRRR

**Comment by hu5970** [comment]
@tanyasmirnova @mdtoyNOAA @joeolson42  Do those surface wind differences make sense after so many changes from MYNN? Thanks.
Right now, we just need to make sure the new code does not produce clear wrong results. We will do full retro soon to check those new code.

**Comment by tanyasmirnova** [comment]
@joeolson42 @mdtoyNOAA @hu5970 @SamuelTrahanNOAA Ming, the new code might increase positive wind bias at the low levels. 
Ming, also, I gave Sam the updates to the diag_table_hrrr, diag_table_hrrr_stoch and diag_table_rap. When he merges my changes, these tables should be updated in the workflow. 
For now you can copy my /mnt/lfs4/BMC/wrfruc/smirnova/RRFS_retro/gsl-srweather-app/regional_workflow/ush/templates//diag_table.FV3_HRRR to your workflow.

**Comment by SamuelTrahanNOAA** [comment]
~~There's a new thompson microphysics feature I had to disable to prevent a crash:~~

EDIT: That is disabled by default in the code. It was the scripts that enabled it.

**Comment by hu5970** [comment]
@tanyasmirnova I will update "regional_workflow/ush/templates//diag_table.FV3_HRRR" to make sure the right diag_table is used in RRFS.

**Comment by hu5970** [comment]
Hi All, Should we try to find a time for a short tag-up to decide how to move forward of this PR before ned of today (early release)?

**Comment by SamuelTrahanNOAA** [comment]
The netcdf I/O still isn't working, but I'm getting close to fixing it.

**Comment by SamuelTrahanNOAA** [comment]
The FV3GFS NetCDF I/O works again.

File: tests/parm/diag_table/diag_table_hrrr [hunk 3]
```diff
@@ -192,7 +192,6 @@
 "gfs_sfc",     "stype",       "sotyp",     "fv3_history2d",  "all",  .false.,  "none",  2
 "gfs_sfc",     "slmsksfc",    "land",      "fv3_history2d",  "all",  .false.,  "none",  2
 "gfs_sfc",     "vfracsfc",    "veg",       "fv3_history2d",  "all",  .false.,  "none",  2
-"gfs_sfc",     "wetness",     "wetness",   "fv3_history2d",  "all",  .false.,  "none",  2
```
Comment (tanyasmirnova) [inline tests/parm/diag_table/diag_table_hrrr]: @SamuelTrahanNOAA Sam, we need this variable in the output to compute fire potential index. Could we revert this change or there is a serious problem with having it in diag_table_hrrr?

Comment (tanyasmirnova) [reply tests/parm/diag_table/diag_table_hrrr]: Also, I made a PR to your repository with changes to diag_table_hrrr, daig_table_hrrr_stoch and diag_table_rap

Comment (SamuelTrahanNOAA) [reply tests/parm/diag_table/diag_table_hrrr]: It was inconsistent. I've added it to both hrrr diag tables now.

**Review by tanyasmirnova — APPROVED** [review]

**Review by tanyasmirnova — APPROVED** [review]

File: tests/tests/rap_noah_sfcdiff_unified_ugwp [hunk 4]
```diff
@@ -93,3 +93,5 @@ export DO_TOFD=.true.
 export GWD_OPT=2
 export DO_GSL_DRAG_SS=.true.
 export DO_GSL_DRAG_TOFD=.true.
+
+export DIAG_FLUX=.true.
```
Comment (tanyasmirnova) [inline tests/tests/rap_noah_sfcdiff_unified_ugwp]: @SamuelTrahanNOAA this suite does not use MYNN surface layer scheme, but uses sfc_diff.f. The diag_flux should be default .false.

File: tests/tests/rap_noah_sfcdiff_unified_ugwp_2threads [hunk 5]
```diff
@@ -98,3 +98,5 @@ export DO_TOFD=.true.
 export GWD_OPT=2
 export DO_GSL_DRAG_SS=.true.
 export DO_GSL_DRAG_TOFD=.true.
+
+export DIAG_FLUX=.true.
```
Comment (tanyasmirnova) [inline tests/tests/rap_noah_sfcdiff_unified_ugwp_2threads]: @SamuelTrahanNOAA same here

File: tests/tests/rap_noah_sfcdiff_unified_ugwp_debug [hunk 6]
```diff
@@ -52,3 +52,5 @@ export DO_TOFD=.true.
 export GWD_OPT=2
 export DO_GSL_DRAG_SS=.true.
 export DO_GSL_DRAG_TOFD=.true.
+
+export DIAG_FLUX=.true.
```
Comment (tanyasmirnova) [inline tests/tests/rap_noah_sfcdiff_unified_ugwp_debug]: same here

File: tests/tests/rap_noah_sfcdiff_unified_ugwp_decomp [hunk 7]
```diff
@@ -95,3 +95,5 @@ export DO_TOFD=.true.
 export GWD_OPT=2
 export DO_GSL_DRAG_SS=.true.
 export DO_GSL_DRAG_TOFD=.true.
+
+export DIAG_FLUX=.true.
```
Comment (tanyasmirnova) [inline tests/tests/rap_noah_sfcdiff_unified_ugwp_decomp]: same here

File: tests/tests/rap_noah_sfcdiff_unified_ugwp_restart [hunk 8]
```diff
@@ -91,3 +91,5 @@ export DO_TOFD=.true.
 export GWD_OPT=2
 export DO_GSL_DRAG_SS=.true.
 export DO_GSL_DRAG_TOFD=.true.
+
+export DIAG_FLUX=.true.
```
Comment (tanyasmirnova) [inline tests/tests/rap_noah_sfcdiff_unified_ugwp_restart]: same here

**Review by tanyasmirnova — COMMENTED** [review]
@SamuelTrahanNOAA diag_flux should be defaul false value for tests using sfcdiff.

File: tests/tests/rap_noah_sfcdiff_unified_ugwp (hunk 4 above)
Comment (SamuelTrahanNOAA) [reply tests/tests/rap_noah_sfcdiff_unified_ugwp]: ~~Corrected.~~ Edit: That one is still in there. I will try changing it back to false and see if the results match.

Comment (SamuelTrahanNOAA) [reply tests/tests/rap_noah_sfcdiff_unified_ugwp]: That test is not run, so it won't pass or fail no matter what I do. I'll remove DIAG_FLUX.
